pkg/controller/clusterstate/clusterstate_controller.go

```diff
@@ -0,0 +1,303 @@
+	}()
+
+	// Fetch the ClusterState instance
+	st := &hivev1.ClusterState{}
+	err := r.Get(context.TODO(), request.NamespacedName, st)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			logger.Info("cluster state not found")
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		logger.WithError(err).Error("Error getting cluster state")
+		return reconcile.Result{}, err
+	}
+	if !st.DeletionTimestamp.IsZero() {
+		logger.Infof("ClusterState resource has been deleted")
+		return reconcile.Result{}, nil
+	}
+
+	// Fetch corresponding ClusterDeployment instance
+	cd := &hivev1.ClusterDeployment{}
+	err = r.Get(context.TODO(), request.NamespacedName, cd)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("cluster deployment not found")
+			return reconcile.Result{}, nil
+		}
+		logger.WithError(err).Error("Error getting cluster deployment")
+		return reconcile.Result{}, err
+	}
+	if !cd.DeletionTimestamp.IsZero() {
+		logger.Infof("ClusterDeployment resource has been deleted")
+		return reconcile.Result{}, nil
+	}
+	if !cd.Status.Installed || len(cd.Status.AdminKubeconfigSecret.Name) == 0 {
+		logger.Infof("ClusterDeployment is not ready")
+		return reconcile.Result{}, nil
+	}
+	kubeconfigSecret := &corev1.Secret{}
+	err = r.Get(context.Background(), types.NamespacedName{Namespace: cd.Namespace, Name: cd.Status.AdminKubeconfigSecret.Name}, kubeconfigSecret)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Warn("admin kubeconfig does not yet exist")
+		} else {
+			return reconcile.Result{}, err
+		}
+	}
+	kubeconfig, err := controllerutils.FixupKubeconfigSecretData(kubeconfigSecret.Data)
+	remoteClient, err := r.remoteClientBuilder(string(kubeconfig), controllerName)
+	if err != nil {
+		logger.WithError(err).Error("error building remote cluster client connection")
+		return reconcile.Result{}, err
+	}
+
+	clusterOperators := &configv1.ClusterOperatorList{}
+	err = remoteClient.List(context.TODO(), clusterOperators)
+	if err != nil {
+		logger.WithError(err).Errorf("failed to list target cluster operators")
+		return reconcile.Result{}, err
+	}
+
+	operatorStates := make([]hivev1.ClusterOperatorState, 0, len(clusterOperators.Items))
+	for _, clusterOperator := range clusterOperators.Items {
+		operatorStates = append(operatorStates, hivev1.ClusterOperatorState{
+			Name:       clusterOperator.Name,
+			Conditions: clusterOperator.Status.Conditions,
+		})
+	}
+	if operatorStatesChanged(logger, st.Status.ClusterOperators, operatorStates) {
+		st.Status.ClusterOperators = operatorStates
+		err = r.updateStatus(r, st)
+		if err != nil {
+			logger.WithError(err).Errorf("failed to update cluster operator state")
+			return reconcile.Result{}, err
+		}
+		logger.Info("clusterState has been updated")
```

Comment:
IMO the code which update or adds cluster operator state should be in a separate function. 
```
operatorStates := make([]hivev1.ClusterOperatorState, 0, len(clusterOperators.Items))
	for _, clusterOperator := range clusterOperators.Items {
		operatorStates = append(operatorStates, hivev1.ClusterOperatorState{
			Name:       clusterOperator.Name,
			Conditions: clusterOperator.Status.Conditions,
		})
	}
	if operatorStatesChanged(logger, st.Status.ClusterOperators, operatorStates) {
		st.Status.ClusterOperators = operatorStates
		err = r.updateStatus(r, st)
		if err != nil {
			logger.WithError(err).Errorf("failed to update cluster operator state")
			return reconcile.Result{}, err
		}
		logger.Info("clusterState has been updated")
```